pkg/az/az.go

```diff
@@ -1,168 +1,62 @@
-	return !avoidZone
-}
-
-// NewZonesToAvoidRule returns a new ZonesToAvoidRule with the supplied
-// zones set to avoid
-func NewZonesToAvoidRule(zoneIDs []string) *ZonesToAvoidRule {
-	z := map[string]struct{}{}
-	for _, zoneID := range zoneIDs {
-		z[zoneID] = struct{}{}
+func GetAvailabilityZones(ec2API ec2iface.EC2API, region string) ([]string, error) {
+	zones, err := getZones(ec2API, region)
+	if err != nil {
+		return nil, err
 	}
-	return &ZonesToAvoidRule{zoneIDs: z}
-}
 
-func makeDefaultZoneUsageRules(region string) []ZoneUsageRule {
-	zoneIDs, ok := zoneIDsToAvoid[region]
-	if !ok {
-		return nil
+	numberOfZones := len(zones)
+	if numberOfZones < api.MinRequiredAvailabilityZones {
+		return nil, fmt.Errorf("only %d zones discovered %v, at least %d are required", numberOfZones, zones, api.MinRequiredAvailabilityZones)
 	}
-	return []ZoneUsageRule{NewZonesToAvoidRule(zoneIDs)}
-}
 
-// AvailabilityZoneSelector used to select availability zones to use
-type AvailabilityZoneSelector struct {
-	ec2api   ec2iface.EC2API
-	region   string
-	strategy SelectionStrategy
-	rules    []ZoneUsageRule
-}
-
-// NewSelectorWithDefaults creates a new AvailabilityZoneSelector with the
-// default selection strategy and usage rules
-func NewSelectorWithDefaults(ec2api ec2iface.EC2API, region string) *AvailabilityZoneSelector {
-	return &AvailabilityZoneSelector{
-		ec2api:   ec2api,
-		strategy: NewRecommendedNumberRandomStrategy(),
-		rules:    makeDefaultZoneUsageRules(region),
-		region:   region,
+	if numberOfZones < api.RecommendedAvailabilityZones {
+		return zones, nil
 	}
-}
 
-// NewSelectorWithMinRequired create a new AvailabilityZoneSelector with the
-// minimum required selection strategy and usage rules
-func NewSelectorWithMinRequired(ec2api ec2iface.EC2API, region string) *AvailabilityZoneSelector {
-	return &AvailabilityZoneSelector{
-		ec2api:   ec2api,
-		strategy: NewMinRequiredNumberRandomStrategy(),
-		rules:    makeDefaultZoneUsageRules(region),
-		region:   region,
-	}
+	return randomSelectionOfZones(region, zones), nil
 }
 
-// SelectZones returns a list fo az zones to use for the supplied region
-func (a *AvailabilityZoneSelector) SelectZones() ([]string, error) {
-	availableZones, err := a.getZonesForRegion()
-	if err != nil {
-		return nil, err
+func randomSelectionOfZones(region string, availableZones []string) []string {
+	var zones []string
+	desiredNumberOfAZs := api.RecommendedAvailabilityZones
+	if region == api.RegionUSEast1 {
+		desiredNumberOfAZs = api.MinRequiredAvailabilityZones
 	}
 
-	usableZones := a.getUsableZones(availableZones)
-
-	return a.strategy.Select(usableZones), nil
-}
-
-func (a *AvailabilityZoneSelector) getUsableZones(availableZones []*ec2.AvailabilityZone) []string {
-	usableZones := []string{}
-	for _, zone := range availableZones {
-		zoneUsable := true
-		for _, rule := range a.rules {
-			if !rule.CanUseZone(zone) {
-				zoneUsable = false
+	for len(zones) < desiredNumberOfAZs {
```

Comment:
Yeah okay, I see what's going on:

```go
	for len(zones) < desiredNumberOfAZs {
		rand := rand.New(rand.NewSource(time.Now().UnixNano()))
		for _, rn := range rand.Perm(len(availableZones)) {
			zones = append(zones, availableZones[rn])
			if len(zones) == desiredNumberOfAZs {
				break
			}
		}
	}
```

This is like this:

```go
		rand := rand.New(rand.NewSource(time.Now().UnixNano()))
		for _, rn := range rand.Perm(len(availableZones)) {
			zones = append(zones, availableZones[rn])
			if len(zones) == desiredNumberOfAZs {
				break
			}
		}
```

If this ever goes wrong than there is something wrong, right? You don't need the outer for. You either exactly the right number of zones or greater number of zones so this would just quit early.

Reply:
![image](https://user-images.githubusercontent.com/19669095/153458736-1b023879-a8b2-4a29-9fac-39c1a9e1d8e1.png)

Reply:
No f*cking way. :D Goddamnit, I thought that's new. :D